Sources/PenguinStructures/AnyValue.swift

```diff
@@ -0,0 +1,218 @@
+// Copyright 2020 Penguin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+fileprivate extension MemoryLayout {
+  /// True iff `T` instances are stored in an existential's inline buffer space.
+  static var fitsExistentialInlineBuffer: Bool {
+    MemoryLayout<(T, Any.Type)>.size <= MemoryLayout<Any>.size
+      && self.alignment <= MemoryLayout<Any>.alignment
+      && _isBitwiseTakable(T.self)
+  }
+}
+
+/// A wrapper for a value of any type, with efficient access and inline storage of bounded size for
+/// small values.
+///
+/// Existential types such as `Any` provide the storage characteristics of `AnyValue`, but many
+/// operations on existentials [optimize poorly](https://bugs.swift.org/browse/SR-13438), so
+/// `AnyValue` may be needed to achieve efficient access.
+///
+/// Using `enums` and no existential, it is possible to build such storage for:
+/// - types that are trivially copiable (such as Int), or
+/// - collections of elements with the inline storage being (roughly) a multiple of the element
+///   size.
+/// You might consider using the `enum` approach where applicable.
+public struct AnyValue {
+  /// The underlying storage of the value.
+  internal var storage: Any
```

Comment:
Although I think I understand what games you're playing with `Box` and `storage` to make up the implementation of `AnyValue`, it's not obvious to a casual reader. I think the documentation on `storage` thus might be somewhat lacking. (Or alternatively, perhaps there's a better place for information on how `AnyValue` is implemented.)

Reply:
Yeah, good catch; the comment is outdated.  I used to be letting the `Any`'s own box do the work, but then couldn't make mutation of a known stored type efficient.  I'll explain that all here.

Reply:
Should be fixed now.